permutations/src/permutation.rs

```diff
@@ -1,28 +1,89 @@
-use std::{collections::HashSet, ops::Mul};
+use std::{collections::HashSet, fmt::Display, ops::Mul, str::FromStr};
 
 use abstraction::{Group, Monoid};
 
+use crate::{disjoint_cycles::DisjointCycles, permutation_error::PermutationError};
+
 /// Mathematical permutation of N elements. An element of the symmetric group S_N
 #[derive(Hash, PartialEq, Eq, Clone, Copy, Debug)]
 pub struct Permutation<const N: usize> {
     values: [usize; N],
 }
 
 impl<const N: usize> Permutation<N> {
-    pub fn new(values: [usize; N]) -> Result<Self, String> {
+    pub fn new(values: [usize; N]) -> Result<Self, PermutationError> {
         let unique_values: HashSet<usize> = HashSet::from_iter(values.iter().cloned());
         if unique_values.len() < values.len() {
-            return Err(String::from("values must not have repeat elements"));
+            return Err(PermutationError::RepeatElement);
         }
 
         for value in values {
-            if value > values.len() {
-                return Err(String::from("values must be less than length of slice"));
+            if value >= N {
+                return Err(PermutationError::ValueOutOfRange(value, N));
             }
         }
 
         Ok(Self { values })
     }
+
+    pub fn from_disjoint_cycles(
+        disjoint_cycles: DisjointCycles<N>,
+    ) -> Result<Self, PermutationError> {
+        let mut combined = [0; N];
+        (0..N).for_each(|i| combined[i] = i);
+
+        let DisjointCycles(cycles) = disjoint_cycles;
+        for cycle in cycles.iter().rev() {
+            for (i, &element) in cycle.iter().enumerate() {
+                combined[element] = cycle[(i + 1) % cycle.len()];
+            }
+        }
+
+        Self::new(combined)
+    }
+
+    /// Compute the cycle decomposition for the permutation.
+    pub fn cycle_decomposition(&self) -> DisjointCycles<N> {
```

Comment:
Test cases:

- permutation without fixed points returns correct permutation
- permutation with fixed points skips those 1-cycles
- cycles are returned in lexicographic order